### DISCUSSION TITLE
Heterogeneous graph link prediction issue

### DISCUSSION THREAD

**Discussion by simonemarasi** [pyg-team/pytorch_geometric#3860] [Q&A] [answered]
Hi, I'm quite new in the PyG world. I need to implement a link prediction task on my custom heterogeneous graph. I generate only some nodes of it in order to simplify the procedure and test step by step the implementation of a GNN for link prediction.
The graph looks as follows:
```
HeteroData(
  images={ x=[56, 1, 1000] },
  attributes={ x=[180, 200] },
  (images, part of, images)={
    edge_index=[2, 46]
  },
  (images, nickname, attributes)={
    edge_index=[2, 38]
  },
  (images, silhouette, attributes)={
    edge_index=[2, 33]
  },
  (images, waistline, attributes)={
    edge_index=[2, 13]
  },
  (images, length, attributes)={
    edge_index=[2, 31]
  })
```

I am trying to apply the link prediction example found [here](https://github.com/pyg-team/pytorch_geometric/blob/8943288ed3f7b70ef6621c4e8b9ec58af3ae24de/examples/hetero/hetero_link_pred.py) but after splitting the dataset into train, val and test I'm now stuck in the part of the model definition. In particular at row 97, when the model tries to initialize the encoder passing the **x_dict** and the **edge_index_dict**
```
with torch.no_grad():
    model.encoder(train_data.x_dict, train_data.edge_index_dict)
```
then it raises the following error: 
`ValueError: `MessagePassing.propagate` only supports `torch.LongTensor` of shape `[2, num_messages]` or `torch_sparse.SparseTensor` for argument `edge_index`.`

It seems reasonable, since the MessagePassing wants tensors of edges, while I am passing the dictionary of the edges.... but with the dataset of the example (and even with other preloaded datasets) the error does not raise... And the HeteroData objects looks of course very similar in the structure. 

Where am I doing wrong? Could anyone help me with some hints? I am stucked here since few days... :(

**Comment by rusty1s** [comment, marked as answer]
Did you forget to call `to_hetero` somewhere? :)

**Reply by simonemarasi** [reply]
> Did you forget to call `to_hetero` somewhere? :)

Thank you for your reply.
The class Model is like the one in the example i linked, namely
```
class Model(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        self.encoder = GNNEncoder(hidden_channels, hidden_channels)
        self.encoder = to_hetero(self.encoder, data.metadata(), aggr='sum')
        self.decoder = EdgeDecoder(hidden_channels)

    def forward(self, x_dict, edge_index_dict, edge_label_index):
        z_dict = self.encoder(x_dict, edge_index_dict)
        return self.decoder(z_dict, edge_label_index)
```

so the `to_hetero()` function in called in the init method, in the definition of the encoder. But, when I call the encoder with the same parameter of the example (the two dicts) it gives me that error. Any idea about that?

**Reply by rusty1s** [reply]
Thanks for your response. Is there any chance you can send me a small reproducible example?

**Reply by simonemarasi** [reply]
> Thanks for your response. Is there any chance you can send me a small reproducible example?

After some troubleshooting I realized that the error was during the creation of the dataset used in the model, so after fixing my dataset it seems to work correctly. Thank you anyway for your help :)
